### DISCUSSION TITLE
HTML judge - use of any_of function

### DISCUSSION THREAD

**Discussion by heremansgeert** [dodona-edu/feedback-and-support#34] [Questions - Vragen] [answered]
# What I want to achieve

I want to check if the submitted HTML contains multiple ```<em>``` tags with specific contents. I am not intrested wether or not this  ```<em>``` is a child of a specific paragraph.

# How I am trying to do it

I have an array with the contents I am looking for. While iterating the array I'm trying to check if an  ```<em>``` exists with this content. I'm trying to use the [any_of](https://github.com/dodona-edu/judge-html/blob/main/docs/pages/utility-functions.md#any_of) function for this.

  ```python
# Import all necessary modules (same for every exercise)
from validators.checks import HtmlSuite, TestSuite, ChecklistItem


def create_suites(content: str) -> list[TestSuite]:
    # Create test suite for HTML with basic validator included
    html = HtmlSuite(content)

    
    #em contents to search
    ems_contents ={
        "the tendency to seek, interpret, and create information that verifies existing beliefs",
        "self-fullfilling prophecy",
        "meest waarschijnlijke reden voor de buitensporige invloed van bevestigende informatie is dat het makkelijker is om er cognitief mee om te gaan",
        "Naar: Verplaetse, J. (2003). Inleiding in de argumentatieleer, Gent: onlinecursus."
    }

    #Get all ems on the page
    foundems = html.all_elements('em')

    # iterate trough the content and check that at least on em is found with the desired content
    for em_content in ems_contents:
         any_of(em.has_content(em_content) for em in foundems)
            
    return [html]
  ```

# The problem I am experiencing

While executing this validator I get an error indicating the [any_of](https://github.com/dodona-edu/judge-html/blob/main/docs/pages/utility-functions.md#any_of)  function does not exists. I am probably forgetting to load a library but din't find which one.

  ```
Traceback (most recent call last):
  File "/mnt/MEnH90L4o0rSguagz0L8Mw/judge/html_judge.py", line 42, in main
    test_suites: List[TestSuite] = evaluator.create_suites(html_content)
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<string>", line 23, in create_suites
NameError: name 'any_of' is not defined
  ```

**Comment by BTWS2** [comment, marked as answer]
You can find the example exercises here: https://github.com/dodona-edu/example-exercises/tree/master/html. Your exercise is similar to https://github.com/dodona-edu/example-exercises/blob/master/html/ugent_faculty_websites/evaluation/evaluator.py.

You need to import the functions you are using from the module e.g. :

```python
from validators.checks import HtmlSuite, CssSuite, TestSuite, ChecklistItem, at_least, all_of, any_of
```

Is your function `any_of` underlined or marked in orange (or similar)? For autocomplete you need to add the folder `validators` with the [checks.pyi](https://github.com/dodona-edu/judge-html/blob/main/validators/checks.pyi) file at the root of your project in which you write the evaluators. [^1]

[^1]: https://github.com/dodona-edu/judge-html?tab=readme-ov-file#quick-start-guide-for-checklist-mode-with-evaluatorpy

**Reply by heremansgeert** [reply]
Thank you once more @BTWS2 for your answer.
